### DISCUSSION TITLE
How to require an option but only if it is specified without a value?

### DISCUSSION THREAD

**Discussion by rcdailey** [Tyrrrz/CliFx#128] [Q&A] [answered]
I have the following option:

```cs
[CommandOption("list-terms", Description = "Test")]
public string? ListTerms { get; set; }
```

There are three variations of the command line:

1. `./mycommand`
1. `./mycommand --list-terms`
1. `./mycommand --list-terms myvalue`

Here is the behavior I want to program for each of the variations above:

1. Command proceeds with behavior unaltered.
2. Command fails and notifies user that `--list-terms` requires a value to be provided.
3. Command proceeds with behavior altered by the specified option and its value.

Right now, 1 & 2 are ambiguous. In other words, I get the same value in the `ListTerms` property when the user doesn't specify the option at all, and when they specify it without a value.

I need a way to require that an option be given an explicit value, but not require the option itself. Is this possible right now? I'm OK implementing the enforcement myself, but that still requires me to disambiguate 1 & 2 above by using the value of the property after command line processing is done by CliFx.

**Comment by Tyrrrz** [comment, marked as answer]
I think you can do it via `BindingValidator` and by setting the validator on the `CommandOption`:

```csharp
[CommandOption("list-terms", Description = "Test", Validators = new[] { typeof(MyValidator)})]
```

If I remember correctly, the validator is only triggered on the value if the option has been provided, so you would simply handle the `null` value which would correspond to case 2 in your situation.

**Reply by rcdailey** [reply]
I don't see documentation or examples for Validators. How would I get access to the validator from within the command? Or rather, how do I know one has been instantiated from the command?

**Reply by rcdailey** [reply]
So I came up with a mechanism to do this, but it's extremely ugly. I think this calls for a built-in feature. Some way to specify a value for an option is required. Here is the solution I came up with.

First, implement a validator:

```cs
internal class ListTermsOptionSpecifiedValidator : BindingValidator<string>
{
    public ListTermsOptionSpecifiedValidator(IDynamicToggles toggles)
    {
        toggles.Flip("list-terms");
    }

    public override BindingValidationError? Validate(string? value)
    {
        return null;
    }
}
```

And the utility service I use to communicate the state from the validator to the command class (fed in via Autofac):

```cs
public interface IDynamicToggles
{
    void Flip(string toggleName);
    bool IsFlipped(string toggleName);
}

internal class DynamicToggles : IDynamicToggles
{
    private readonly HashSet<string> _toggles = new();
    public void Flip(string toggleName) => _toggles.Add(toggleName);
    public bool IsFlipped(string toggleName) => _toggles.Contains(toggleName);
}
```

And in my Command class:

```cs
[Command("sonarr", Description = "Perform operations on a Sonarr instance")]
public class SonarrCommand : ServiceCommand
{
    private readonly IDynamicToggles _toggles;

    [CommandOption("list-terms", Validators = new[] {typeof(ListTermsOptionSpecifiedValidator)}, Description =
        "For the given Release Profile Trash ID, list terms in it that can be filtered in YAML format. " +
        "Note that not every release profile has terms that may be filtered.")]
    public string? ListTerms { get; [UsedImplicitly] set; }

    public SonarrCommand(IDynamicToggles toggles)
    {
        _toggles = toggles;
    }

    public async ValueTask ExecuteAsync(IConsole console)
    {
        if (_toggles.IsFlipped("list-terms"))
        {
            if (!string.IsNullOrEmpty(ListTerms))
            {
                // do something meaningful
            }
            else
            {
                throw new CommandException(
                    "The --list-terms option was specified without a Release Profile Trash ID specified");
            }

            return;
        }
    }
```

I mean, this works as you say it should, but I just think it would be so much simpler if you had a new attribute property to specify the value as required. Something like `DefaultAllowed = false`. Then it would require a user to specify the value and not allow any default to be provided.

**Reply by Tyrrrz** [reply]
I don't understand exactly what you did there but what I meant was:

```csharp
    [CommandOption("list-terms", Validators = new[] {typeof(ListTermsOptionSpecifiedValidator)}, Description =
        "For the given Release Profile Trash ID, list terms in it that can be filtered in YAML format. " +
        "Note that not every release profile has terms that may be filtered.")]
    public string? ListTerms { get; [UsedImplicitly] set; }
```

```csharp
internal class ListTermsOptionSpecifiedValidator : BindingValidator<string>
{
    public override BindingValidationError? Validate(string? value)
    {
        if (value is null)
        	// Option has been provided, but with null value
           return Error("Bla bla bla");

        return Ok();
    }
}
```

This will print something like:

```powershell
Option --list-terms has been provided with an invalid value.
Error(s):
- Bla bla bla
```

There are no docs on `BindingValidator`s, but you can look at some examples in tests:

https://github.com/Tyrrrz/CliFx/blob/ed3054c855043f5e1c3768047d42f2ef62b15ece/CliFx.Tests/ConversionSpecs.cs#L906-L948

**Reply by rcdailey** [reply]
I prefer to keep business rules in my command class not in a validator. Either way, I feel like validators are being abused here which is why I suggested this be a new feature.

**Reply by Tyrrrz** [reply]
Okay then you can just do this:

```csharp
    [CommandOption("list-terms", Description =
        "For the given Release Profile Trash ID, list terms in it that can be filtered in YAML format. " +
        "Note that not every release profile has terms that may be filtered.")]
    public string? ListTerms { get; [UsedImplicitly] set; } = "";
```

Then you will have three states:

1. `./mycommand` → `ListTerms` is empty
2. `./mycommand --list-terms` → `ListTerms` is `null`
3. `./mycommand --list-terms myvalue` → `ListTerms` is `myvalue`

There's also another edge case:

4. `./mycommand --list-terms ""` → `ListTerms` is empty

if you want to differentiate between 4 and 1, then you can set `ListTerms` default value to some magic string then you're sure isn't going to be used as a value. Case 4 will always result in an empty value.

**Reply by rcdailey** [reply]
Ok this is a little better, thank you.

If you're not objected to it, I'd be happy to contribute a PR to make an attribute property for this. What do you think? If I were to spend time on this, how would you like it to be done? I certainly wouldn't want to start working on a feature you won't accept or that isn't otherwise acceptable because it doesn't function as you like or consistent with the design principles of the library.

Let me know if and how you'd like to discuss it.

**Reply by Tyrrrz** [reply]
I'd say let's hold it off until it becomes something more people request. As of now, both workarounds above are easier than designing and implementing a new feature.
